pandas/core/nanops.py

```diff
@@ -829,9 +830,42 @@ def _checked_add_with_arr(arr, b):
 
     Raises
     ------
-    OverflowError if any x + y exceeds the maximum int64 value.
+    OverflowError if any x + y exceeds the maximum or minimum int64 value.
     """
-    if (np.iinfo(np.int64).max - b < arr).any():
-        raise OverflowError("Python int too large to "
-                            "convert to C long")
+    # For performance reasons, we broadcast 'b' to the new array 'b2'
+    # so that it has the same size as 'arr'.
+    if np.__version__ >= LooseVersion('1.10.0'):
+        b2 = np.broadcast_to(b, arr.shape)
+    else:
+        if lib.isscalar(b):
+            b2 = np.empty(arr.shape)
+            b2.fill(b)
+        else:
+            b2 = b.copy()
+
+    # For each element in 'arr' and its corresponding element in 'b2',
+    # we check the sign of the element in 'b2'. If it is positive, we
+    # then check whether its sum with the element in 'arr' exceeds
+    # np.iinfo(np.int64).max. If so, we have an overflow error. If it
+    # it is negative, we then check whether its sum with the element in
+    # 'arr' exceeds np.iinfo(np.int64).min. If so, we have an overflow
+    # error as well.
+    #
+    # gh-14324: The implementation below takes advantage of NumPy
+    # broadcasting to check all of these elements at once. Although
+    # it is somewhat convoluted, it has much better performance than
+    # say iterating through all elements in both 'arr' and 'b2'.
+    mask1 = b2 > 0
+    mask2 = b2 < 0
+
+    if not mask1.any():
+        to_raise = (np.iinfo(np.int64).min - b2 > arr).any()
+    elif not mask2.any():
+        to_raise = (np.iinfo(np.int64).max - b2 < arr).any()
+    else:
+        to_raise = ((np.iinfo(np.int64).max - b2[mask1] < arr[mask1]).any() or
+                    (np.iinfo(np.int64).min - b2[mask1] > arr[mask1]).any())
```

Comment:
I suppose one of these two lines should use mask2 instead of mask1 (probably an error in my example code)